### PR TITLE
feat: support bm25 logs mixcompaction

### PR DESCRIPTION
**Pull request by aoiasd**
relate: https://github.com/milvus-io/milvus/issues/35853

### CODE REVIEW

File: internal/storage/binlog_iterator.go [hunk 1]
```diff
@@ -46,6 +46,7 @@ type Iterator interface {
 type Value struct {
 	ID        int64
 	PK        PrimaryKey
+	BM25Row   map[int64][]byte
```
Comment (zhengbuqian) [inline internal/storage/binlog_iterator.go:49]: is this new member necessary?

In `segment_writer.go` we already have `w.bm25Stats` initialized with field id as key, we should be able to:

```go
for fieldID, stats := range w.bm25Stats {
    stats.AppendBytes(v.Value.(map[storage.FieldID]interface{})[fieldID].([]byte)
}
```

in `(w *SegmentWriter) Write`. also we can avoid changes in `serde_events.go`.

File: internal/indexnode/task_stats.go [hunk 2]
```diff
@@ -564,6 +581,44 @@ func statSerializeWrite(ctx context.Context, io io.BinlogIO, startID int64, writ
...
+	stats, err := writer.GetBm25Stats()
+	if err != nil {
+		return 0, nil, err
+	}
+
+	kvs := make(map[string][]byte)
+	binlogs := []*datapb.FieldBinlog{}
+	cnt := int64(0)
+	for fieldID, blob := range stats {
+		key, _ := binlog.BuildLogPath(storage.BM25Binlog, writer.GetCollectionID(), writer.GetPartitionID(), writer.GetSegmentID(), fieldID, startID+cnt)
```
Comment (zhengbuqian) [inline internal/indexnode/task_stats.go:596]: do not ignore error

File: internal/datanode/compaction/segment_writer.go [hunk 3]
```diff
@@ -350,6 +366,10 @@ func (w *SegmentWriter) Write(v *storage.Value) error {
 	}
 
 	w.pkstats.Update(v.PK)
+	for fieldId, bytes := range v.BM25Row {
```
Comment (zhengbuqian) [inline internal/datanode/compaction/segment_writer.go:369]: can we do?
```go
	for fieldID, stats := range w.bm25Stats {
		stats.AppendBytes(v.Value.(map[storage.FieldID]interface{})[fieldID].([]byte))
	}
```

File: internal/datanode/compaction/compactor_common.go [hunk 4]
```diff
@@ -229,3 +229,45 @@ func mergeFieldBinlogs(base, paths map[typeutil.UniqueID]*datapb.FieldBinlog) {
 		base[fID].Binlogs = append(base[fID].Binlogs, fpath.GetBinlogs()...)
 	}
 }
+
+func bmSerializeWrite(ctx context.Context, io io.BinlogIO, allocator allocator.Interface, writer *SegmentWriter) ([]*datapb.FieldBinlog, error) {
```
Comment (zhengbuqian) [inline internal/datanode/compaction/compactor_common.go:233]: this method is mostly the same as `bm25SerializerWrite` in `indexnode/task_stats.go`, avoid duplicate code

File: internal/datacoord/task_stats.go [hunk 5]
```diff
@@ -178,7 +178,9 @@ func (st *statsTask) PreCheck(ctx context.Context, dependency *taskScheduler) bo
 		return false
 	}
 
-	start, end, err := dependency.allocator.AllocN(segment.getSegmentSize() / Params.DataNodeCfg.BinLogMaxSize.GetAsInt64() * int64(len(collInfo.Schema.GetFields())) * 2)
+	bm25LogNum := int64(len(collInfo.Schema.GetFunctions()))
+	binlogNum := (segment.getSegmentSize()/Params.DataNodeCfg.BinLogMaxSize.GetAsInt64() + 1) * int64(len(collInfo.Schema.GetFields())) * 2
```
Comment (czs007) [inline internal/datacoord/task_stats.go:182]: 2 may be not enough if there are multi type stats in field.
for now ,we can change it to 100.

File: internal/indexnode/task_stats.go [hunk 6]
```diff
@@ -564,6 +579,44 @@ func statSerializeWrite(ctx context.Context, io io.BinlogIO, startID int64, writ
 	return binlogNum, statFieldLog, nil
 }
 
+func bm25SerializerWrite(ctx context.Context, io io.BinlogIO, startID int64, writer *compaction.SegmentWriter, finalRowCount int64) (int64, []*datapb.FieldBinlog, error) {
```
Comment (czs007) [inline internal/indexnode/task_stats.go:582]: in datanode/compaction/compactor_common.go there is a function named bm25SerializeWrite

Comment (aoiasd) [reply internal/indexnode/task_stats.go:582]: They have different way to get logID.
Compaction and index task could use same code to do serialize, but not this pr. We need optimize binlog and stats log together.

File: internal/datanode/compaction/clustering_compactor.go [hunk 7]
```diff
@@ -1270,6 +1315,48 @@ func (t *clusteringCompactionTask) checkBuffersAfterCompaction() error {
 	return nil
 }
 
+func (t *clusteringCompactionTask) generateBM25Stats(ctx context.Context, segmentID int64, statsMap map[int64]*storage.BM25Stats) ([]*datapb.FieldBinlog, error) {
```
Comment (czs007) [inline internal/datanode/compaction/clustering_compactor.go:1318]: duplicate code with bm25SerializeWriter

Comment (czs007) [reply internal/datanode/compaction/clustering_compactor.go:1318]: refine this in next pr

**Pull request merged**